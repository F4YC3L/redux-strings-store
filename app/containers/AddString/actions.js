/*
 * Home Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  CHANGE_NEW_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_ERROR,
  ADD_STRING,
} from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {string} newString The new text of the input field
 *
 * @return {object} An action object with a type of CHANGE_NEW_STRING
 */
export function changeNewString(newString) {
  return {
    type: CHANGE_NEW_STRING,
    newString,
  };
}

/**
 * add string, this action starts the request saga
 *
 * @return {object} An action object with a type of ADD_STRING
 */
export function addString() {
  return {
    type: ADD_STRING,
  };
}

/**
 * add string, this action starts the request saga
 *
 * @return {object} An action object with a type of ADD_STRING_SUCCESS
 */
export function stringAdded() {
  return {
    type: ADD_STRING_SUCCESS,
  };
}

/**
 * Dispatched when loading the data fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of ADD_STRING_ERROR passing the error
 */
export function stringAddingError(error) {
  return {
    type: ADD_STRING_ERROR,
    error,
  };
}
