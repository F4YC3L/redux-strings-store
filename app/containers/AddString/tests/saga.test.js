/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';
import { dataLoaded } from 'containers/App/actions';
import { stringAddingError } from '../actions';
import { ADD_STRING } from '../constants';
import addStringApi, { addString } from '../saga';

const data = {
  id: 'a5bf98c-7cfe-4531-b321-c9e99c4ce658',
  text: 'JavaScript',
};

/* eslint-disable redux-saga/yield-effects */
describe('addString Saga', () => {
  let addStringGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    addStringGenerator = addString();

    const selectDescriptor = addStringGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = addStringGenerator.next(data).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the dataLoaded action if it requests the data successfully', () => {
    const response = [
      {
        id: 'a5bf98c-7cfe-4531-b321-c9e99c4ce658',
        text: 'JavaScript',
      },
    ];
    const putDescriptor = addStringGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(dataLoaded(response)));
  });

  it('should call the stringAddingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = addStringGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(stringAddingError(response)));
  });
});

describe('addStringApi Saga', () => {
  const addStringApiSaga = addStringApi();

  it('should start task to watch for ADD_STRING action', () => {
    const takeLatestDescriptor = addStringApiSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(ADD_STRING, addString));
  });
});
