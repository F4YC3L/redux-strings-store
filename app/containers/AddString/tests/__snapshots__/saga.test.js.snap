// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addString Saga should call the stringAddingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addString Saga should call the stringAddingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api/add",
      Object {
        "body": "{\\"id\\":\\"722515b9-e4bc-4d28-8e17-1b9ccea7d780\\",\\"text\\":{\\"id\\":\\"a5bf98c-7cfe-4531-b321-c9e99c4ce658\\",\\"text\\":\\"JavaScript\\"}}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`addString Saga should dispatch the dataLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addString Saga should dispatch the dataLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api/add",
      Object {
        "body": "{\\"id\\":\\"907c54a7-71d2-48aa-8f4e-cd7bfca930ff\\",\\"text\\":{\\"id\\":\\"a5bf98c-7cfe-4531-b321-c9e99c4ce658\\",\\"text\\":\\"JavaScript\\"}}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
